libs/server/Lua/LuaRunner.Functions.cs

```diff
@@ -2773,6 +2774,813 @@ static bool TryDecodeLargeMap(LuaRunner self, ref ReadOnlySpan<byte> data, out i
+
+                if (h.Endian == Native.Endian) // header endian == native endian
+                {
+                    // If header endian matches native endian, pick the method that matches native order
+                    if (BitConverter.IsLittleEndian)
+                        BinaryPrimitives.WriteDoubleLittleEndian(into[0..], value);
+                    else
+                        BinaryPrimitives.WriteDoubleBigEndian(into[0..], value);
+                }
+                else
+                {
+                    // Endian mismatch: pick the opposite method to get the bytes in header's endian format
+                    if (BitConverter.IsLittleEndian)
+                        BinaryPrimitives.WriteDoubleBigEndian(into[0..], value);
+                    else
+                        BinaryPrimitives.WriteDoubleLittleEndian(into[0..], value);
+                }
+
+                self.scratchBufferBuilder.MoveOffset(8);
+                self.state.Remove(stackIndex);
+
+                constStrRegisteryIndex = -1;
+                return true;
+            }
+            
+            static bool TryEncodeBytes(LuaRunner self, int stackIndex, char opt, ref int size, out int constStrRegisteryIndex)
+            {
+                Debug.Assert(self.state.Type(stackIndex) == LuaType.String, "Expected string");
+
+                self.state.KnownStringToBuffer(stackIndex, out var data);
+
+                int l = data.Length;
+
+                if (size == 0)
+                {
+                    size = l;
+                }
+
+                if (l < size)
+                {
+                    // string too short, expected at least {size} bytes but got {l}
+                    constStrRegisteryIndex = self.LuaWrappedError(1, self.constStrs.ErrBadArg);
+                    return false;
+                }
+
+                var into = self.scratchBufferBuilder.ViewRemainingArgSlice(data.Length).Span;
+                data.CopyTo(into[0..]);
+                self.scratchBufferBuilder.MoveOffset(data.Length);
+
+                if (opt is 's')
+                {
+                    AddAlignmentPadding(self);
+                    size++;
+                }
+
+                self.state.Remove(stackIndex);
+
+                constStrRegisteryIndex = -1;
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Entry point for struct.unpack from a Lua script.
+        /// </summary>
+        internal int StructUnpack(nint luaStatePtr)
+        {
+            state.CallFromLuaEntered(luaStatePtr);
+
+            var numLuaArgs = state.StackTop;
+
+            if (numLuaArgs == 0 || !TryGetFormat(this, 1, out var format))
+            {
+                return LuaWrappedError(0, constStrs.BadArgPack);
+            }
+
+            // Use stackIndex 1 become when try get format, it also removes format from the lus stack
+            state.KnownStringToBuffer(1, out var data);
```

Comment:
How do we know this is a string?

Reply:
I assume it's a string from here 
https://github.com/redis/redis/blob/ecd5e639ed4d35f71c3e67889879aee0e8fb3919/deps/lua/src/lua_struct.c#L297

Added an assertion to check the type, not sure if I need to throw a lua error here?